src/coreclr-embedding-test/src/ScriptingManager.cpp

```diff
@@ -0,0 +1,149 @@
+#define DOTNET_RUNTIME_DELEGATE "hostfxr_get_runtime_delegate"
+#define DOTNET_RUN_FUNCTION "hostfxr_run_app"
+#define DOTNET_CLOSE_FUNCTION "hostfxr_close"
+#define DOTNET_ERROR_WRITER "hostfxr_set_error_writer"
+
+#if WIN32
+#define HOST_FXR_PATH "hostfxr.dll"
+#elif __linux__
+#define HOST_FXR_PATH "host/fxr/8.0.0/libhostfxr.so"
+#elif __APPLE__
+#define HOST_FXR_PATH "host/fxr/8.0.0/libhostfxr.dylib"
+#endif
+
+template <class T>
+T* ScriptingManager::InvokeCSharp(const char* targetAssembly, const char* targetNamespace, const char* targetClass, const char* fnName, void* args...) {
+	InvokeCSharp(targetAssembly, targetNamespace, targetClass, fnName, args);
+	return nullptr;
+}
+
+
+void ScriptingManager::InvokeCSharp(const char* targetAssembly, const char* targetNamespace, const char* targetClass, const char* fnName, void* args...) {
+	//TODO: Consider caching the functions in memory to a map so that we don't have to constantly load them every time
+	//I'll make this more readable in the future, and we'll probably accept some heap allocs, for now I want to test if it works being performant
+	//Concatenate them as {targetNamespace.targetClass}, {assemblyName}
+	const int MAX_ASSEMBLY_DECL = 2048; //Dedicate 2MBs to the target
+	std::string fullClassPath;
+	fullClassPath.reserve(MAX_ASSEMBLY_DECL);
+	fullClassPath += targetNamespace;
+	fullClassPath += '.';
+	fullClassPath += targetClass;
+	fullClassPath += ", ";
+	fullClassPath += targetAssembly;
+	
+	//Allocate memory to concatenate the string
+	//Get the correct type of function pointer
+	using test_delegate_fn = void (*)(void**);
+	test_delegate_fn test_delegate;
+	int rc = this->function_getter_fptr(
+							  fullClassPath.c_str(),
+							  fnName,
+							  UNMANAGEDCALLERSONLY_METHOD,
+							  nullptr,
+							  nullptr,
+							  reinterpret_cast<void**>(&test_delegate));
+	if (rc != 0) {
+		fputs("Error invoking CSharp method with name: ", stderr);
+		fputs(fnName, stderr);
+		fputs(". Please verify the signature\n", stderr);
+		return;
+	}
+	test_delegate(&args);
+}
+
+
+template <class T>
+T ScriptingManager::LoadSymbol(void *sharedLibrary, const char *name) {
+	return reinterpret_cast<T>(dlsym(sharedLibrary, name));
+}
+
+ScriptingManager::ScriptingManager(const char* dotnetPath) {
+	//Get env for dotnet here and use that as our path
+	//auto sharedLibrary = dlopen("/usr/share/dotnet/host/fxr/8.0.0/libhostfxr.so", RTLD_LAZY);
+	std::filesystem::path targetPath = dotnetPath;
+	targetPath /= HOST_FXR_PATH;
+	void* sharedLibrary = dlopen(targetPath.c_str(), RTLD_LAZY);
+	if (!sharedLibrary)
+	{
+		fputs("Failed to load ", stderr);
+		fputs(targetPath.c_str(), stderr);
+		fputs("\n", stderr);
+		return;
+	}
+	//TODO: See which of these can stop being cached and just pass them as params to the initdotnetcore
+	this->cmd_line_fptr = LoadSymbol<hostfxr_initialize_for_dotnet_command_line_fn>(sharedLibrary, DOTNET_CMD);
```

Comment:
Should we use `this` everywhere?
I ask this to include that in the clang-tidy file. :)

Reply:
I personally like to, since it makes the reference object more clear